asf_core_data/pipeline/mcs/process/mcs_epc_joining.py

```diff
@@ -0,0 +1,392 @@
+    verbose=True,
+):
+    """Join prepared MCS and EPC data.
+
+    Args:
+        hps (Dataframe): Dataframe with standardised_postcode,
+        numeric_tokens and standardised_address fields.
+        epcs (Dataframe): Dataframe with standardised_postcode,
+        numeric_tokens and standardised_address fields.
+        all_records (bool, optional): Whether all top matches should be kept,
+        or just one. Keeping all records allows for comparison of
+        property characteristics over time. Defaults to True.
+        drop_epc_address (bool, optional): Whether or not to drop addresses
+        from the EPC records. Useful to keep for determining whether
+        matches are sensible. Defaults to True.
+        verbose (bool, optional): Whether or not to print diagnostic information
+        about the matching, e.g. number of matched records. Defaults to True.
+
+    Returns:
+        Dataframe: Merged MCS and EPC records.
+    """
+
+    print("Forming a matching...")
+    matching = form_matching(df1=hps, df2=epcs)
+
+    # First ensure that all matches are above the matching parameter
+    good_matches = matching[
+        (matching["numerics"] == 1) & (matching["address_score"] >= matching_parameter)
+    ].reset_index()
+
+    if all_records:
+        top_matches = (
+            good_matches.groupby("level_0")
+            # Get all level_1 indices of rows in which address_score is maximal
+            .apply(
+                lambda df: df.loc[
+                    df["address_score"] == df["address_score"].max(), "level_1"
+                ]
+            )
+            .droplevel(1)
+            .reset_index()
+            .rename(columns={"index": "level_0"})
+        )
+
+    else:
+        # Get the level_1 index of the first occurrence at which address_score is maximal
+        top_match_indices = good_matches.groupby("level_0")["address_score"].idxmax()
+
+        # Filter the matches df to just the top matches
+        top_matches = good_matches.loc[top_match_indices].drop(
+            columns=["numerics", "address_score"]
+        )
+
+    print("Joining the data...")
+    merged = (
+        hps.reset_index()
+        # Join MCS records to the index-matching df on MCS index
+        .merge(top_matches, how="left", left_on="index", right_on="level_0")
+        # Then join this merged df to EPC records on EPC index
+        .merge(epcs.reset_index(), how="left", left_on="level_1", right_on="index")
+        # Drop any duplicated or unnecessary columns
+        .drop(
+            columns=[
+                "standardised_postcode_x",
+                "standardised_address_x",
+                "numeric_tokens_x",
+                "ADDRESS1",
+                "ADDRESS2",
+                "POSTTOWN",
+                "POSTCODE",
+                "index_y",
+                "postcode_y",
+                "standardised_postcode_y",
+                "numeric_tokens_y",
+                "level_0",
+            ],
```

Comment:
To avoid/reduce the x_y  columns, you could remove some of them before joining, e.g.

`.merge(epcs.reset_index()[`_only columns you really need_`]`

or `hps.drop(columns=[`_columns you like numeric tokens and others you drop later_`]`

Personally, I try to avoid creating columns with x and y tags if at all possible because I'm always worried I will confuse them one day.

Reply:
Nice idea, done